### ISSUE TITLE
Add configuration option to be able to click URLs without holding CTRL

### ISSUE THREAD

**Issue by pragma-** [mintty/mintty#1169]
I desperately want to be able to configure mintty to open URLs with a simple left-click without needing to hold CTRL. A configuration option to allow this would be very welcome!

**Comment by pragma-** [comment]
Additionally, a related option to always underline URL-like text while mousing over them may be worth consideration.

**Comment by pragma-** [comment]
Food for thought: perhaps this could be implemented simply as a boolean switch to invert the initial state of the CTRL key?

**Comment by BrianInglis** [comment]
> I desperately want to be able to configure mintty to open URLs with a simple left-click without needing to hold CTRL. A configuration option to allow this would be very welcome!

You just double-click, and drag if less than a line, or triple-click if a whole line, to select, then right-click and select Open. 
The colon ":" after URI scheme is a word break, so most of the time you need to drag over that.

**Comment by BrianInglis** [comment]
> Additionally, it may be worth considering a related boolean option to always underline URL-like text or not while hovering over such text with the mouse, regardless of the state of the CTRL/modifier key.

That would likely slow mouse interaction, to search back and forth to match the start and end of general URIs, unless it could be done incrementally and in parallel with normal receive operation, but that could slow all terminal operation.

**Comment by pragma-** [comment]
When I tested holding CTRL down and moving the mouse around, causing various words to be underlined, there was no significant difference in CPU usage compared to moving the mouse around without holding CTRL.

I do not have right-click bound to opening the context menu where "Open" lives. I instead have it bound to paste because that is what my muscle memory is most familiar with. Nor do I want to drag a selection to highlight a URL when holding CTRL produces the desired effect of finding and underlining complete URLs with minimal effort.

Consequently, I would much appreciate toggle-able option that allows my instance of mintty to behave as if CTRL is always held down. This does not degrade performance on my system and it provides me with the behavior I desperately wish for. If others are concerned about the performance impact, they can leave this option at its default disabled value.

**Comment by mintty** [comment]
> as if CTRL is always held down.

I assume you mean only in connection with the mouse - otherwise all letters would become control characters.

An experimental implementation works - kind of - but then click-and-drag selection does not work anymore.
This may be a fiddly feature...

**Comment by pragma-** [comment]
@mintty: Right, strictly in connection with the mouse.

I'm trying to reproduce clicking-and-dragging the selection but I do not understand the process for this. I searched briefly through https://mintty.github.io/mintty.1.html for instances of `drag` and `selection` but did not hit upon the relevant section. How do you click-and-drag the selection? I tried several combinations of modifier keys -- CTRL + ALT moved the window, ALT made the selection square, etc.

Well, by "behaving as if CTRL was always held down" I more rather meant to refer to mintty's internal state flag specifically related to URL-underlining and URL-clicking with the mouse. Perhaps the quick-and-easy solution is more fiddly than anticipated, but surely there's a flag that can be isolated or added that doesn't affect the other selection modifier flags?

Thank you for considering this issue.

**Comment by pragma-** [comment]
Oh, it's SHIFT! The SHIFT key controls click and drag!

**Comment by pragma-** [comment]
@mintty: I have been heavily editing my comments on this issue. I'm unsure if you're reading these comments via e-mail notifications or via the GitHub website or some command-line tools or whatnot. I do not know if you're aware of my edits. So I'm going to copy and paste, as its own comment, an edit that I do not want you to miss:

>Well, by "behaving as if CTRL was always held down" I more rather meant to refer to mintty's internal state flag specifically related to URL-underlining and URL-clicking with the mouse (if there is such a flag). Perhaps the quick-and-easy solution is more fiddly than anticipated, but surely there's a flag that can be isolated or added that doesn't affect any modifier keys or selection modifier flags?

Apologies for all the edits and comments.

**Comment by mintty** [comment]
You clidk the mouse and drag it to make a selection, like in any other program.
This normally works without Shift, unless you are running a program that uses application mouse mode, to capture mouse events itself. This would be disabled by general Ctrl remapping, also e.g. Ctrl+mouse-wheel (for zooming) etc.

**Comment by mintty** [comment]
The opening function can be confined to the case that we were hovering, i.e. a link was highlighted/underlined.
However, may words on the screen will be available as links, because mintty also supports filenames to be opened.
On the other hand, you asked explicitly about URLs, so I think it would be appropriate to apply the non-click opening function to URLs only.

**Comment by pragma-** [comment]
Oh, normal selection! I thought you meant clicking on the selection itself and dragging it to resize it smaller or to extend it further, which I've just learned can be accomplished by holding the Shift key as long as you are -- as you stated -- not running a program that uses application mouse mode, such as `set mouse=a` in vim. When using application mouse mode, the Shift key instead seems to temporarily disable application mouse mode while the Shift key is being held down. That's neat.

I hadn't realized mintty supported clicking normal words to attempt opening them as local files. I do see how every word gets underlined as I hover the mouse over them. While I initially explicitly asked about URLs, I don't really mind if non-URLs continue to remain underlined and clickable as well.

**Comment by mintty** [comment]
Well, the intent of my question was rather reverse: If you allow all potential filenames for opening, there will hardly remain any clickable text to make a selection... I don't think that would be a very useful mode.

**Comment by pragma-** [comment]
I'm not sure I understand the question. I am unfamiliar with how mintty handles mouse clicks under the hood. I would expect that click-and-release-without-moving-the-mouse would invoke the "Open" behavior (for URLs or filenames), whereas click-and-hold-while-moving-the-mouse would invoke the "Selection" behavior.

**Comment by pragma-** [comment]
Also, when allowing to click-to-open for URLs and filenames, I would expect double-click to retain its select-a-whole-word functionality and triple-click to retain its select-a-whole-line functionality. I do not understand why this behavior would need to change.

**Comment by mintty** [comment]
> click-and-release-without-moving-the-mouse would invoke the "Open" behavior (for URLs or filenames), 
> click-and-hold-while-moving-the-mouse would invoke the "Selection" behavior.

That would be intuitive but it's not covered by the current implementation scheme. That's why the feature is going to be fiddly.

**Comment by pragma-** [comment]
Actually, having thought about it a bit more, I see what you're saying. Web browsers don't allow people to click and drag while over  click to select the text (which I think is weird). People must hold ALT to change to selection mode while clicking on a link, in order to drag a selection. Perhaps something similar can be done? But if limiting to just URLs makes this simpler to implement, I'm all for it.

**Comment by pragma-** [comment]
Oh, right, I forget that people like to be able to click-and-drag hyperlinks to their bookmarks or to another app to open them there and whatnot. Makes sense, actually. So yeah, if limiting the click-to-open to just URLs makes selecting text less fiddly, I'm all for it.

**Comment by pragma-** [comment]
On the other hand, if the hold-CTRL behavior is inverted, shouldn't people now be able to hold CTRL to select text while hovering over a URL or filename?

**Comment by pragma-** [comment]
Keep in mind that this behavior would be an option that isn't enabled by default. Hopefully people would read the documentation for this option to understand its fiddly caveats?

**Comment by pragma-** [comment]
Of course, implementing this in the least fiddly way with a minimal of surprising caveats would be desirable! :-)

**Comment by pragma-** [comment]
@mintty, has there been any progress on the "experimental build that kind of works" that was mentioned above? May I download this build somewhere? Perhaps the fiddly issues mentioned may be tolerable for my use-case. I would like to try it out, if possible.

**Comment by mintty** [comment]
Wait a few days, I'm working on a sound solution.

**Comment by pragma-** [comment]
I manually compiled this a couple of days ago. I've been testing it out. It works really well! Thank you so much.

While playing around with this, I did have a thought/suggestion. What do you think about making it so that if someone clicks and holds down the button while dragging the mouse then it'll go into text-selection mode as long as the mouse button is held down? This way we could single-click to open URLs and files, but hold-click-and-drag to select text.

**Comment by mintty** [comment]
Thanks for testing, and for the suggestion. From an interactive design perspective, yes. Implementation will be fiddly though, and I'm not sure I'd like to spend the effort. It's not only a quick solution (which might be found) but a significant amount of consideration and testing about possibly affected cases would be involved.

**Comment by mintty** [comment]
Implementation of text selection was a bit fiddly, but not too much :) Please test.
I change option name LinkMod to OpeningMod.

**Comment by pragma-** [comment]
Nice! I've only tested with `OpeningMod=off` and the selection seems to work well now! Thank you.
